### ISSUE TITLE
Tracking issue - GraphQL performance (Slow or poor relation performance)

### ISSUE THREAD

**Issue by derrickmehaffy** [strapi/strapi#8552]
## **Describe the bug**

This is a tracking issue for all the various issues related to GraphQL performance as compared to REST or just in general poor relational performance

## **Context**

Currently GraphQL is performance very un-optimized on model entries that have quite a lot of relations or a very complex relational structure. This is not what we intended as GraphQL itself and the query system implemented should provide a much faster response as compared to something like REST, as the goal of GraphQL is to be faster however with a higher degree of control.

## **Current behavior**

(Taken from the OP of: https://github.com/strapi/strapi/issues/5182 but see linked issues for more current behavior)

> I've noticed some slow performance whenever the amount of relations started growing of an entry. I thought the fix mentioned in #4080 already fixed this, but after some testing, I've come up with the following results (same query/same output fields):
> 
> REST: 110-120ms
> GraphQL: 900-910ms
> 
> That's a big difference. After reading the original issue, I started to dig through the graphql plugin and started fiddling with the populate settings on line 155 in services/Loader.js (as already mentioned in the original post). Setting the populate to a hard coded [] results in a huge boost:
> 
> GraphQL: 120-140ms
> 
> The funny thing here is, the fields that originally are populated in the params, are still queryable after setting the populate to [].
> However, this does not seem to have effect when deep filtering. I not sure why, but i'll try to figure that out sometime this week.

## **Expected behavior**

GraphQL performance should be significantly faster than REST in many different use cases. Or at the very least, performance wise, in line with our REST implementation.

## **Related Issues, Feature requests, Discussions, and Forum threads**

Please note, the below are not the limit and additional entries may be linked to this issue, you can find them referenced in the comments.

- https://github.com/strapi/strapi/issues/5182
- https://github.com/strapi/strapi/issues/4080
- https://github.com/strapi/strapi/issues/8535
- https://github.com/strapi/strapi/issues/5363
- https://github.com/strapi/strapi/issues/3552

More will be added as needed

## **Possible workarounds depending on the use case**

None currently known other than to avoid usage of GraphQL in favor of REST

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/trying-to-understand-how-graphql-plugin-works/1078/2

**Comment by xarvel** [comment]
Any news ?

**Comment by derrickmehaffy** [comment]
@xarvel it's going to take some time for us to address this probably and most likley won't happen until Q1 2021

**Comment by jwerle** [comment]
@xarvel to improve performance, we've introduced [graphql-jit](https://github.com/zalando-incubator/graphql-jit) as an executor for Apollo, rewrote how the dataloaders initialize and work in a request contexts, removed `"dominant": true` from all model attributes, overload how components are loaded in resolvers, and introduced a cache with Apollo. We still see higher latency with GraphQL compared to REST, but the average request latency has dropped with these changes. We believe the root of our issues comes from the **has many and belongs to many** relationship between models, but we are still actively auditing.

This was a lot of guessing with trial/error overtime :man_shrugging:

**Comment by derrickmehaffy** [comment]
> We believe the root of our issues comes from the has many and belongs to many relationship between models, but we are still actively auditing.

This is a major part of the problem yes, I **very strongly** suggest you avoid this relation unless it's absolute requirement for your project (I'd say about 75% of the time I see this relation being used, it's not being used in the correct context, but we have no information to help users design an efficient schema and we plan to in the upcoming doc refresh)

Generally speaking, most of the performance issues related to relation population in general (or fields) where we are requesting everything from the database, then cleaning up the data before passing it to the user. Instead Strapi needs to only request the fields required.

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community**. There might be relevant details there:

https://forum.strapi.io/t/caching-graphql-query-results-with-strapi/1407/1

**Comment by derrickmehaffy** [comment]
This issue has been mentioned on **Strapi Community Forum**. There might be relevant details there:

https://forum.strapi.io/t/slow-query-due-to-relation/2300/2

**Comment by wyamout** [comment]
6 month in development, after migrating tons of data...Boom! a slap in the face with the slow queries.

**Comment by abdonrd** [comment]
Is there an estimate of when this problem will be addressed?

**Comment by wyamout** [comment]
@abdonrd I heard Q3

**Comment by derrickmehaffy** [comment]
Yes, we will likely be taking a look in Q3 with the database refresh: https://portal.productboard.com/strapi/1-roadmap/c/73-database-layer-v4

**Comment by abdonrd** [comment]
With some of the latest versions, the performance has gotten even worse. 😢

**Comment by abdonrd** [comment]
I just tested that from `v3.4.1` to `v3.4.2`, querying relationships becomes much slower.

**Comment by alexandrebodin** [comment]
What kind of relations is that ?

**Comment by abdonrd** [comment]
> What kind of relations is that ?

{
  "kind": "collectionType",
  "collectionName": "battle_stages",
  "info": {
    "name": "Battle Stage",
    "description": ""
  },
  "options": {
    "increments": true,
    "timestamps": true,
    "draftAndPublish": false
  },
  "attributes": {
    // ...
    "battles": {
      "autoPopulate": false,
      "via": "stage",
      "collection": "battle"
    }
  }
}

**Comment by abdonrd** [comment]
> What kind of relations is that ?

```json
{
  "kind": "collectionType",
  "collectionName": "battles",
  "info": {
    "name": "Battle",
    "description": ""
  },
  "options": {
    "increments": true,
    "timestamps": true,
    "draftAndPublish": true
  },
  "attributes": {
    ...
    "winners": {
      "autoPopulate": false,
      "collection": "profile",
      "via": "battlesWon",
      "dominant": true
    }
  }
}
```

**Comment by alexandrebodin** [comment]
Ok so a simple many many thanks. I'll run some benchmarking in the upcoming days.

**Comment by abdonrd** [comment]
Thanks @alexandrebodin! 🙏

We have had to downgrade to `v3.4.1` for now, because otherwise GraphQL plugin is unusable.

**Comment by alexandrebodin** [comment]
can you give me a brief approximate of the volumes of data ?

**Comment by abdonrd** [comment]
> can you give me a brief approximate of the volumes of data ?

±6.000 battles and ±3.500 profiles.

**Comment by alexandrebodin** [comment]
@abdonrd can you tell me approx the volume of the relation (count of links)

**Comment by abdonrd** [comment]
> @abdonrd can you tell me approx the volume of the relation (count of links)

Sorry, what do you mean?

**Comment by alexandrebodin** [comment]
can you test this @abdonrd  https://github.com/strapi/strapi/pull/9286

**Comment by Jbmanllr** [comment]
So in short, strapi graphql is unusable?

**Comment by abdonrd** [comment]
> So in short, strapi graphql is unusable?

No. It is working much better in the latest version!

**Comment by Jbmanllr** [comment]
> > So in short, strapi graphql is unusable?
> 
> No. It is working much better in the latest version!

Hi, im not very backend savy, could you explain what is going on with strapi graphql and what is the difference between using it or a custom graphql server?

**Comment by joelbqz** [comment]
Another possible workaround would be to use Strapi with PostgreSQL DB then connect a Hasura Instance, Hasura is pretty much schema-agnostic and performant, you just need to spend a couple of minutes setting up relationships but you will have a blazing fast GraphQL engine ready for production and a good CMS(Strapi) for non-technical users, I love Strapi as CMS and I have used it as Node framework to build 2 apps with it, but as API provider, it's not very efficient.

**Comment by derrickmehaffy** [comment]
> > > So in short, strapi graphql is unusable?
> > 
> > 
> > No. It is working much better in the latest version!
> 
> Hi, im not very backend savy, could you explain what is going on with strapi graphql and what is the difference between using it or a custom graphql server?

The short answer is our graphql plugin is currently functioning as a Koa middleware so most of the performance issues are due to relation resolving as gql inside of Strapi bypasses your typical REST code to resolving and populating relations.

We plan to refactor the GQL plugin in v4

**Comment by abdonrd** [comment]
> We plan to refactor the GQL plugin in v4

That could take Strapi to the next level!

**Comment by dodikuswono** [comment]
> 
> 
> Another possible workaround would be to use Strapi with PostgreSQL DB then connect a Hasura Instance, Hasura is pretty much schema-agnostic and performant, you just need to spend a couple of minutes setting up relationships but you will have a blazing fast GraphQL engine ready for production and a good CMS(Strapi) for non-technical users, I love Strapi as CMS and I have used it as Node framework to build 2 apps with it, but as API provider, it's not very efficient.

Interesting idea. Anyone exploring this approach and run on production ?
Does subscription work ?

**Comment by Boytobeaman** [comment]
> > We plan to refactor the GQL plugin in v4
> 
> That could take Strapi to the next level!

any time plan for this, buddy? I have a very slow response when using graphql request

**Comment by derrickmehaffy** [comment]
@Boytobeaman we will begin v4 later this year, but no static ETA dates yet.

**Comment by seloner** [comment]
Hello guys I wanted to use graphql in my latest project . Do you recommend going with REST instead ? Is the issue with graphql huge ?

**Comment by ralphsomeday** [comment]
Had a big performance issue with GraphQL but it was due to relations being fetched when not needed so I quickly fixed it and graphql is working like a charm. You have to be very careful with fetching the right data when needed and not just use strapi find method without the right populate argument.

**Comment by seloner** [comment]
> Had a big performance issue with GraphQL but it was due to relations being fetched when not needed so I quickly fixed it and graphql is working like a charm. You have to be very careful with fetching the right data when needed and not just use strapi find method without the right populate argument.

Thanks for your help . What kind of relation caused the issue ?

**Comment by ralphsomeday** [comment]
> > Had a big performance issue with GraphQL but it was due to relations being fetched when not needed so I quickly fixed it and graphql is working like a charm. You have to be very careful with fetching the right data when needed and not just use strapi find method without the right populate argument.
> 
> Thanks for your help . What kind of relation caused the issue ?

In my case it was a one to many relation where the data was growing on a daily basis on that relation and I was doing a find on the object that has a many relation with another object and because i did not fill in the populate argument then every day the results from that find query was growing until it started hitting the server CPU and slowing down the entire platform. So use the find method with the populate argument so that you make sure you fetch the needed data within you policies, controllers, services, middlewares, plugins.

**Comment by seloner** [comment]
> > > Had a big performance issue with GraphQL but it was due to relations being fetched when not needed so I quickly fixed it and graphql is working like a charm. You have to be very careful with fetching the right data when needed and not just use strapi find method without the right populate argument.
> > 
> > 
> > Thanks for your help . What kind of relation caused the issue ?
> 
> In my case it was a one to many relation where the data was growing on a daily basis on that relation and I was doing a find on the object that has a many relation with another object and because i did not fill in the populate argument then every day the results from that find query was growing until it started hitting the server CPU and slowing down the entire platform. So use the find method with the populate argument so that you make sure you fetch the needed data within you policies, controllers, services, middlewares, plugins.

Are you talking about things you are doing server side?I am talking from client side.Can you use populate using the API endpoints?I am building an e-commerce  app .  Do you think that I should switch to rest instead of graphql?

**Comment by sofyan-ahmad** [comment]
Is there any update on this?

At the moment this is the workaround that I do to improve the graphql performance.
https://sofyan-ahmad.medium.com/strapi-io-boost-graphql-performance-with-redis-cache-e0dac5051472

TLDR; Basically what I did:
- Implement cache on Apollo
- To fix the permission issue, I call permission.js before returning any values from the cache

**Comment by tention** [comment]
> > > Had a big performance issue with GraphQL but it was due to relations being fetched when not needed so I quickly fixed it and graphql is working like a charm. You have to be very careful with fetching the right data when needed and not just use strapi find method without the right populate argument.
> > 
> > 
> > Thanks for your help . What kind of relation caused the issue ?
> 
> In my case it was a one to many relation where the data was growing on a daily basis on that relation and I was doing a find on the object that has a many relation with another object and because i did not fill in the populate argument then every day the results from that find query was growing until it started hitting the server CPU and slowing down the entire platform. So use the find method with the populate argument so that you make sure you fetch the needed data within you policies, controllers, services, middlewares, plugins.

@ralphsomeday Do you mind to share how you limit the relation? I tried setting populate argument, but Graphql seems to ignore it and continue sending all relationships over. This is what I have tried, but Graphql still return all relationship other than the 2 fields.

  result = await strapi.services.articles.findOne({ uid: uid }, ['field1', 'field2']);

**Comment by ralphsomeday** [comment]
You just have to do it in your graphql query. Do not request relations you do not need and in the controller set the populate parameter to only return the data you need and that should already improve perfs.

**Comment by tention** [comment]
Thanks @ralphsomeday. Limiting the fields on graphql query doesn't really help in my case, because when I checked via console.log, the full set of data is still being queried from database, and somehow being filtered after that.

I have also tried configuring the populate as below, with populate being the array of data to populate, but still the full list of relationships are being queried as gathered by console.log output.

```
  async find(populate) {
    const results = await strapi.query('restaurant').find({ _limit: 1 }, populate);
    console.log(results);
  },
```

**Comment by ralphsomeday** [comment]
Can you share your model and the query that you are calling?

**Comment by tention** [comment]
I believed it's due to the same issue with this, where the populate is not working properly and caused all the data to be returned.

https://github.com/strapi/strapi/issues/9970

**Comment by ivanji** [comment]
Have you guys considered also supporting Persisted queries? https://www.apollographql.com/blog/announcement/platform/improve-graphql-performance-with-automatic-persisted-queries/ - It might help in some aspects with performance. If you so how could we take advantage of it?

**Comment by derrickmehaffy** [comment]
Fixed in v4, marking as closed

**Comment by jwerle** [comment]
great work strapi team on v4!

**Comment by jorisw** [comment]
Could you link to the PR that fixed this? Curious what's changed.

**Comment by shierro** [comment]
Good stuff!

Would you mind adding more details here please? maybe benchmark results as
to why the team thinks that the GraphQL performance is resolved for good.
Some other questions in mind: Does it work well on both MongoDB &
PostgreSQL databases? Is the GraphQL now as good as the REST API or maybe
better?

**Comment by derrickmehaffy** [comment]
> Could you link to the PR that fixed this? Curious what's changed.

There is no one PR for it, effectively rewriting the entire GraphQL plugin and the complete database redesign is what fixed it.

> Some other questions in mind: Does it work well on both MongoDB & PostgreSQL databases?

We no longer support MongoDB on v4

> Is the GraphQL now as good as the REST API or maybe better?

REST is still far better than GQL but the performance on GQL is still much better than it was in v3 (mainly database query layer related)

> Would you mind adding more details here please?

I haven't done extensive benchmarks, but in limited testing I'd estimate normal complex queries are executing in about 10ms and for extremely complex queries (like 11 layers deep) about 30ms running locally (no remote network latency)
